### PR TITLE
Fix struct flattening to add a validity column only when the input column has null element

### LINKED ISSUE #8187: [BUG] mismatched columns on struct lower bound
**Describe the bug**
In `lower_bound` and `upper_bound` for structs it calls `structs::detail::flatten_nested_columns` for both `t` and `values` but it does this independently for both of them. That means if we end up with a `t` that  has null structs but a `values` that does not, or vise versa we can get errors.

**Steps/Code to reproduce bug**
```diff
diff --git a/cpp/tests/search/search_struct_test.cpp b/cpp/tests/search/search_struct_test.cpp
index 50c326269a..c127f6c8b6 100644
--- a/cpp/tests/search/search_struct_test.cpp
+++ b/cpp/tests/search/search_struct_test.cpp
@@ -152,8 +152,8 @@ TYPED_TEST(TypedStructSearchTest, SimpleInputWithNullsTests)
   auto const structs_values = structs_col{{child_col_values}, null_at(3)}.release();
 
   // Sorted asc, nulls first
-  auto child_col_t = col_wrapper{{XXX, null, 0, 1, 2, 2, 2, 2, 3, 3, 4}, null_at(1)};
-  auto structs_t   = structs_col{{child_col_t}, null_at(0)}.release();
+  auto child_col_t = col_wrapper{{0, 0, 0, 1, 2, 2, 2, 2, 3, 3, 4}};
+  auto structs_t   = structs_col{{child_col_t}}.release();
 
   auto results =
     search_bounds(structs_t, structs_values, {cudf::order::ASCENDING}, {cudf::null_order::BEFORE});
```

This causes the error to happen. I have not updated the expected result to what they should be, but it does expose the error.

```
[ RUN      ] TypedStructSearchTest/17.SimpleInputWithNullsTests
unknown file: Failure
C++ exception with description "cuDF failure at: ../include/cudf/table/row_operators.cuh:361: Mismatched number of columns." thrown in the test body.
[  FAILED  ] TypedStructSearchTest/17.SimpleInputWithNullsTests, where TypeParam = cudf::detail::timestamp<cuda::std::__3::chrono::duration<long, cuda::std::__3::ratio<1l, 1000l> > > (0 ms)
```

### PR DESCRIPTION
**Pull request by ttnghia**
Currently, struct flattening adds a validity column when the input column has a null mask (by calling to `nullable()`). In the situation when comparing two structs columns having no null but one column has a null mask, flattening them will result in two tables with different numbers of columns.

This PR fix that problem by using `has_nulls()` instead of `nullable()`. As a result, the validity column will be added to the flattening result only when the input structs column has null.

Note that when comparing two structs columns in which one column has null while the other doesn't, we must check for (nested) null existence and pass in `column_nullability::FORCE` for flattening both columns. This makes sure the flattening results are tables having the same number of columns.

Closes #8187.

### CODE REVIEW

**Comment by abellina** [comment]
Locally this fixes the issue reported by @gerashegalov

**Review by davidwendt — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
LGTM!

File: cpp/src/structs/utilities.cpp [hunk 1]
```diff
@@ -93,11 +93,16 @@ struct flattened_table {
...
                              null_order col_null_order)
   {
-    if (nullability == column_nullability::FORCE || col.nullable()) {
-      // nullable columns could be required for comparisions such as joins
+    // Even if it is not required to extract the bitmask to a separate column,
+    // we should always do that if the structs column has any null element.
+    // In addition, we should check for null by calling to `has_nulls()`, not `nullable()`.
+    // This is because when comparing structs columns, if one column has bitmask while the other
+    // does not (and both columns do not have any null element) then flattening them using
+    // `nullable()` will result in tables with different number of columns.
```
Comment (jrhemstad) [inline cpp/src/structs/utilities.cpp:101]: This explanation doesn't make sense to me. Don't you have the same problem if one column has `has_nulls() == true` and the other has `has_nulls() == false`?

The solution should be to use `column_nullability::FORCE` on both the LHS and RHS like @hyperbolic2346 did https://github.com/rapidsai/cudf/pull/7963

Comment (ttnghia) [reply cpp/src/structs/utilities.cpp:101]: When one column has null while the other doesn't, during calling to `search()`, `column_nullability::FORCE` will be passed in.

Comment (jrhemstad) [reply cpp/src/structs/utilities.cpp:102]: Keep in mind that `has_nulls()` isn't always free, so the performance trade off isn't totally clear cut.

**Comment by harrism** [comment]
@gpucibot merge

**Comment by gerashegalov** [comment]
Can you add a test case (ok if in another PR) so we don't defer to spark-rapids to catch these issues

**Comment by ttnghia** [comment]
> Can you add a test case (ok if in another PR) so we don't defer to spark-rapids to catch these issues

Sure. But can that test be targeted to branch 21.08?

**Comment by gerashegalov** [comment]
> Sure. But since we are about to do code freeze, can that test be targeted to branch 21.08?

I think it's reasonable since @abellina confirmed the fix with a Spark test run

**Pull request merged**